Automattic/woocommerce-payments
# Fix invalid value error for paymentMethodMessaging in product details page

Fixes #6819 (reopened)

This issue has been reopened. We received reports of errors in the plugin in the product details page after the previous fix was released. This time, errors seem to happen due to merchants customizing this page by removing the quantity input.

#### Changes proposed in this Pull Request

- Improved the method of fetching the product quantity and handling cases where the input element might be missing.
- Introduced safeguards against potential undefined values for productVariations and related fields.
- Established 'USD' as the fallback currency in case it's not defined in productVariations.
- Added more strict checks for potentially missing DOM elements to prevent script errors.









#### Testing instructions




To replicate the issue we need to remove the quantity input from the product details page. This can be achieved by several means. We got reports of merchants using a page builder. I was able to replicate the issue locally, using this hook `woocommerce_is_sold_individually`. In the functions.php file of the Storefront theme add the following:

```
function custom_woocommerce_remove_quantity_input() {
    return true;
}
add_filter( 'woocommerce_is_sold_individually', 'custom_woocommerce_remove_quantity_input' );
```
* Enable Affirm or Afterpay in your local store.
* Navigate to the details page of a product with variations.
* Check the Affirm and Afterpay messaging is displayed when the page loads.
* Check there are no errors in the browser console related to `paymentMethodMessaging`.
* Check you are able to select variations with no errors happening.
* Check you are able to add the product to the cart without issues.
* Navigate to the details page of a product **without** variations.
* Check the same steps as before.
* Enable the quantity selector back (you can return false in the `custom_woocommerce_remove_quantity_input`)
* Repeat the same steps to check you are able to add the product to the cart.

-------------------


**Post merge**
